source/darwin/securityframework_rsa.c

```diff
@@ -0,0 +1,545 @@
+        CFRelease(rsa_key->pub_key_ref);
+    }
+
+    if (rsa_key->priv_key_ref) {
+        CFRelease(rsa_key->priv_key_ref);
+    }
+
+    if (rsa_key->cf_allocator) {
+        aws_wrapped_cf_allocator_destroy(rsa_key->cf_allocator);
+    }
+
+    aws_mem_release(key_pair->allocator, rsa_key);
+}
+
+/*
+ * return either pointer to a key algo (global consts) or NULL in case algo is
+ * not supported.
+ */
+static int s_map_rsa_encryption_algo_to_sec(enum aws_rsa_encryption_algorithm algorithm, SecKeyAlgorithm *out) {
+
+    switch (algorithm) {
+        case AWS_CAL_RSA_ENCRYPTION_PKCS1_5:
+            *out = kSecKeyAlgorithmRSAEncryptionPKCS1;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_ENCRYPTION_OAEP_SHA256:
+            *out = kSecKeyAlgorithmRSAEncryptionOAEPSHA256;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_ENCRYPTION_OAEP_SHA512:
+            *out = kSecKeyAlgorithmRSAEncryptionOAEPSHA512;
+            return AWS_OP_SUCCESS;
+    }
+
+    return AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+}
+
+/*
+ * return either pointer to a key algo (global consts) or NULL in case algo is
+ * not supported.
+ */
+static int s_map_rsa_signing_algo_to_sec(enum aws_rsa_signing_algorithm algorithm, SecKeyAlgorithm *out) {
+
+    switch (algorithm) {
+        case AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256:
+            *out = kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256;
+            return AWS_OP_SUCCESS;
+        case AWS_CAL_RSA_SIGNATURE_PSS_SHA256:
+            if (__builtin_available(macos 10.13, ios 11.0, tvos 11.0, watchos 4.0, *)) {
+                *out = kSecKeyAlgorithmRSASignatureDigestPSSSHA256;
+                return AWS_OP_SUCCESS;
+            } else {
+                return AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+            }
+    }
+
+    return AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM;
+}
+
+int s_rsa_encrypt(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor plaintext,
+    struct aws_byte_buf *out) {
+    struct sec_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    if (key_pair_impl->pub_key_ref == NULL) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "RSA Key Pair is missing Public Key required for encrypt operation.");
+        return aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
+    }
+
+    SecKeyAlgorithm alg;
+    if (s_map_rsa_encryption_algo_to_sec(algorithm, &alg)) {
+        return AWS_OP_ERR;
+    }
+
+    if (!SecKeyIsAlgorithmSupported(key_pair_impl->pub_key_ref, kSecKeyOperationTypeEncrypt, alg)) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Algo is not supported for this operation");
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    CFDataRef plaintext_ref = CFDataCreateWithBytesNoCopy(NULL, plaintext.ptr, plaintext.len, kCFAllocatorNull);
+    AWS_FATAL_ASSERT(
+        plaintext_ref && "No allocations should have happened here, this function shouldn't be able to fail.");
```

Comment:
i copied this pattern from ecc impl without much thinking and yes this is wrong and we should be using our cf_allocator instead of NULL, which will fallback to default allocator.

Comment in general: I borrowed some questionable coding practices from ecc implementation that are called out in other comments, especially around error handling. Once this review is done im planning to do a quick pass through ecc implementation and fix it up a bit